### PR TITLE
[awsemfexporter] Restructure Metric Translator Logic

### PR DESCRIPTION
**Pull request by kohrapha**
**Description:**
This PR restructures the code for building CW Metrics from pdata.Metric to remove duplicated logic. To do this, we had to define the following interfaces:
1. `DataPoints` that wraps `pdata.(Int|Double|DoubleHistogram)DataPointSlice`
2. `DataPoint` that wraps pdata.(Int|Double|DoubleHistogram)DataPoint

However, since the `At(i)` method for `DataPoints` requires returning a `DataPoint`, which is not the same as what `pdata.(Int|Double|DoubleHistogram)DataPoint` returns, we have to define a light conversion struct that provides the correct method signature.

The following functions have also been renamed to communicate its intended output better:
- `getMeasurements` -> `getCWMetrics`
- `buildCWMetricFromDP` and `buildCWMetricFromHistogram` are combined into `buildCWMetric`

This code change should not have any visible API changes. It's merely a reorganization of the code to clean it up and remove code duplication.

**Testing:**
This CR was run against the unit tests successfully.

**Performance**
I ran some benchmarks to compare the performance of this new code change against the old code, and got the following results:

```
benchmark                                                old ns/op     new ns/op     delta
BenchmarkTranslateOtToCWMetricWithInstrLibrary-12        9588          9182          -4.23%
BenchmarkTranslateOtToCWMetricWithoutInstrLibrary-12     8836          8475          -4.09%
BenchmarkTranslateOtToCWMetricWithNamespace-12           17.4          16.5          -5.17%
BenchmarkTranslateCWMetricToEMF-12                       23867         22283         -6.64%

benchmark                                                old allocs     new allocs     delta
BenchmarkTranslateOtToCWMetricWithInstrLibrary-12        131            122            -6.87%
BenchmarkTranslateOtToCWMetricWithoutInstrLibrary-12     116            112            -3.45%
BenchmarkTranslateOtToCWMetricWithNamespace-12           0              0              +0.00%
BenchmarkTranslateCWMetricToEMF-12                       45             45             +0.00%

benchmark                                                old bytes     new bytes     delta
BenchmarkTranslateOtToCWMetricWithInstrLibrary-12        7462          7190          -3.65%
BenchmarkTranslateOtToCWMetricWithoutInstrLibrary-12     6853          6869          +0.23%
BenchmarkTranslateOtToCWMetricWithNamespace-12           0             0             +0.00%
BenchmarkTranslateCWMetricToEMF-12                       84394         84391         -0.00%
```

We note that, not only is there no negative impact on the performance, we see a minor improvement across the board. This is great news because we managed not to impact performance while improving code readability for developers.

### CODE REVIEW

**Comment by kohrapha** [comment]
Tagging @mxiamxia, @hdj630, and @shaochengwang for visibility.

File: exporter/awsemfexporter/metric_translator.go [hunk 1]
```diff
@@ -150,217 +191,127 @@ func TranslateCWMetricToEMF(cwMetricLists []*CWMetrics) []*LogEvent {
...
+	fields = make(map[string]interface{})
+	dimensionKV := dp.LabelsMap()
 
+	dimensionSlice := make([]string, dimensionKV.Len(), dimensionKV.Len()+1)
+	idx := 0
 	dimensionKV.ForEach(func(k string, v pdata.StringValue) {
-		fieldsPairs[k] = v.Value()
-		dimensionSlice = append(dimensionSlice, k)
+		fields[k] = v.Value()
+		dimensionSlice[idx] = k
```
Comment (anuraaga) [inline exporter/awsemfexporter/metric_translator.go:297]: Not sure but FWIU, it's still idiomatic to stick with append even when preallocating the array in Go.

File: exporter/awsemfexporter/metric_translator.go [hunk 2]
```diff
@@ -150,217 +191,127 @@ func TranslateCWMetricToEMF(cwMetricLists []*CWMetrics) []*LogEvent {
...
-			if cwMetric != nil {
-				result = append(result, cwMetric)
-			}
-		}
+		dps = DoubleDataPointSlice{metric.DoubleSum().DataPoints()}
 	case pdata.MetricDataTypeDoubleHistogram:
-		dps := metric.DoubleHistogram().DataPoints()
-		if dps.Len() == 0 {
-			return result
+		dps = DoubleHistogramDataPointSlice{metric.DoubleHistogram().DataPoints()}
```
Comment (anuraaga) [inline exporter/awsemfexporter/metric_translator.go:217]: Seems like a great improvement - separate from this PR would be interesting if this is useful to other exporters and could be in collector core somewhere

**Review by anuraaga — APPROVED** [review]
Just small nit but looks nice, thanks

File: exporter/awsemfexporter/metric_translator.go (hunk 2 above)
Comment (kohrapha) [reply exporter/awsemfexporter/metric_translator.go:217]: Yea, would be useful to see if we can move this interface to collector core for exporters to use.

File: exporter/awsemfexporter/metric_translator.go (hunk 1 above)
Comment (kohrapha) [reply exporter/awsemfexporter/metric_translator.go:297]: Oh interesting, I can change it if you'd like. I just did some research and the general opinion on SO was to use indexing if you know the length ahead of time.

Comment (bogdandrutu) [reply exporter/awsemfexporter/metric_translator.go:297]: If you do this, you avoid checking for capacity all the time, so it is indeed a small improvement.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**